### PR TITLE
feat(Line - Contact): Add bulk action to contact edit

### PR DESCRIPTION
**Pull request by feoche**
## Title of the Pull Requests
feat(Line - Contact): Add bulk action to contact edit

### Description of the Change
- Add bulk action to contact edit
- Add an extra function bound on bulk button clicking

### CODE REVIEW

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 1]
```diff
@@ -46,6 +46,7 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
         }).catch(function (err) {
             return new ToastError(err);
         }).finally(function () {
+            self.getFilteredDirectoryForm();
```
Comment (JeremyDec) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: no need to call method here

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 2]
```diff
@@ -104,6 +104,57 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
...
     };
 
+    self.getFilteredDirectoryForm = function () {
+        OvhApiTelephony.Lexi().schema({
+            billingAccount: $stateParams.billingAccount,
+            serviceName: $stateParams.serviceName
+        }, {}).$promise.then(function (infos) {
+            const properties = infos.models && infos.models && infos.models["telephony.DirectoryInfo"] && infos.models["telephony.DirectoryInfo"].properties;
+            self.filteredDirectoryForm = _.pick(self.directoryForm, (v, k) => _.get(properties, [k, "readOnly"]) === 0);
+            console.log(self.filteredDirectoryForm);
```
Comment (JeremyDec) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: console.log

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 3]
```diff
@@ -104,6 +104,56 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
...
     };
 
+    self.getFilteredDirectoryForm = function () {
+        OvhApiTelephony.Lexi().schema({
+            billingAccount: $stateParams.billingAccount,
+            serviceName: $stateParams.serviceName
+        }, {}).$promise.then(function (infos) {
+            const properties = infos.models && infos.models && infos.models["telephony.DirectoryInfo"] && infos.models["telephony.DirectoryInfo"].properties;
+            self.filteredDirectoryForm = _.pick(self.directoryForm, (v, k) => _.get(properties, [k, "readOnly"]) === 0);
+            return self.filteredDirectoryForm;
```
Comment (JeremyDec) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: useless

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 4]
```diff
@@ -104,6 +104,55 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
         });
     };
 
+    self.getFilteredDirectoryForm = function () {
+        OvhApiTelephony.Lexi().schema({
+            billingAccount: $stateParams.billingAccount,
+            serviceName: $stateParams.serviceName
+        }, {}).$promise.then(function (infos) {
+            const properties = infos.models && infos.models && infos.models["telephony.DirectoryInfo"] && infos.models["telephony.DirectoryInfo"].properties;
```
Comment (antleblanc) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]:
```js
const properties = infos.models && infos.models && infos.models["telephony.DirectoryInfo"] && infos.models["telephony.DirectoryInfo"].properties;
```
could be written like so?

```js
const properties = _.get(infos, "models.telephony.DirectoryInfo.properties");
```

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html [hunk 5]
```diff
@@ -713,29 +713,43 @@
...
+                                        data-ng-if="ServiceContactCtrl.isEditing"
+                                        data-ng-disabled="!ServiceContactCtrl.hasChanges() || ServiceContactCtrl.isUpdating || contactForm.$invalid">
+                                    <oui-loader class="mr-2"
+                                                data-ng-if="ServiceContactCtrl.isUpdating"
+                                                data-size="s">
+                                    </oui-loader>
+                                    <span data-translate="telephony_service_contact_apply"></span>
+                                </button>
+                                <button type="button"
+                                        class="btn btn-default"
```
Comment (antleblanc) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html]: I think it should be `btn-default` => `btn-link`.

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 6]
```diff
@@ -104,6 +104,55 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
         });
     };
 
+    self.getFilteredDirectoryForm = function () {
+        OvhApiTelephony.Lexi().schema({
+            billingAccount: $stateParams.billingAccount,
```
Comment (Jisay) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: ```billingAccount``` and ```serviceName``` params are not needed. You can remove them

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 7]
```diff
@@ -104,6 +104,55 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
...
+            serviceName: $stateParams.serviceName
+        }, {}).$promise.then(function (infos) {
+            const properties = infos.models && infos.models && infos.models["telephony.DirectoryInfo"] && infos.models["telephony.DirectoryInfo"].properties;
+            self.filteredDirectoryForm = _.pick(self.directoryForm, (v, k) => _.get(properties, [k, "readOnly"]) === 0);
```
Comment (Jisay) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: can we avoid (for the moment) to have a mix of old js and es6 ? Not that I don't like es6 but for a better readability, we should use the same anotations.

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html [hunk 8]
```diff
@@ -713,29 +713,43 @@
...
+                    <div class="form-group mt-4">
+                        <div class="row">
+                            <div class="col-md-6">
+                                <hr/>
+                                <telephony-bulk-action data-billing-account="{{ ServiceContactCtrl.bulkDatas.billingAccount }}"
+                                                       data-service-name="{{ ServiceContactCtrl.bulkDatas.serviceName }}"
+                                                       data-custom-class="link d-block pl-0 mb-3"
+                                                       data-bulk-infos="ServiceContactCtrl.bulkDatas.infos"
+                                                       data-get-bulk-params="ServiceContactCtrl.getBulkParams"
+                                                       data-on-open="ServiceContactCtrl.applyChanges() && ServiceContactCtrl.getFilteredDirectoryForm()"
```
Comment (Jisay) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html]: why don't you use ```filterServices``` callback ?
And why do you apply changes before modal open ?

Comment (feoche) [reply client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html]: We need filtered data based on an external GET method that should be triggered on modal loading.

Comment (Jisay) [reply client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html]: Why not add this call on controller init and wait that the promise is resolved before enabling bulk button and filter services ?

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js [hunk 9]
```diff
@@ -46,6 +47,13 @@ angular.module("managerApp").controller("TelecomTelephonyServiceContactCtrl", fu
         }).catch(function (err) {
             return new ToastError(err);
         }).finally(function () {
+            OvhApiTelephony.Lexi().schema({}, {}).$promise.then(function (infos) {
```
Comment (Jisay) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.controller.js]: Use a ```$q.all``` and return it in the init method. For the time this request take, you can block bulk during the whole initialization.

File: client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html [hunk 10]
```diff
@@ -713,29 +713,44 @@
...
+                        <div class="row">
+                            <div class="col-md-6">
+                                <hr/>
+                                <telephony-bulk-action data-billing-account="{{ ServiceContactCtrl.bulkDatas.billingAccount }}"
+                                                       data-service-name="{{ ServiceContactCtrl.bulkDatas.serviceName }}"
+                                                       data-custom-class="link d-block pl-0 mb-3"
+                                                       data-ng-disabled="ServiceContactCtrl.isBulkLoading"
+                                                       data-bulk-infos="ServiceContactCtrl.bulkDatas.infos"
+                                                       data-get-bulk-params="ServiceContactCtrl.getBulkParams"
+                                                       data-on-open="ServiceContactCtrl.applyChanges()"
```
Comment (Jisay) [inline client/app/telecom/telephony/service/contact/telecom-telephony-service-contact.html]: I don't understand : the directory changes must be applied to current service on bulk modal open ?

**Pull request merged**